content/blog/2022-04-21-debug-your-lamda-functions-from-your-ide-using-localstack/index.md

```diff
@@ -0,0 +1,242 @@
+
+This blog will look over how you can debug your Python, and JVM Lambda functions from VS Code. You will also see how to configure your local IDE for local debugging and make the best use of LocalStack’s functionality for Lambda debugging and testing.
+
+## Configuring LocalStack for Python debugging
+
+In this blog, we already presume that you have installed LocalStack and tried running it locally at least once. If not, you can install LocalStack using the following command via `pip`:
+
+```sh
+pip install localstack
+```
+
+It will install the `localstack-cli`, which runs the Docker image that hosts the LocalStack runtime. You can start LocalStack in a detached mode by running the following command:
+
+```sh
+localstack start -d
+```
+
+To ensure that we can use the remote debugging feature, we need to start LocalStack using specific configurations. These are detailed on our [Lambda configuration page](https://docs.localstack.cloud/localstack/configuration/#lambda). Push the following command onto your terminal:
+
+```sh
+LAMBDA_REMOTE_DOCKER=0 \ 
+LAMBDA_DOCKER_FLAGS='-p 19891:19891' \ 
+DEBUG=1 localstack start -d
+```
+
+The `LAMBDA_REMOTE_DOCKER` is configured as false to ensure that the Lambda volume mounts work while we are mounting a temporary folder on the host. The `LAMBDA_DOCKER_FLAGS` defines a Docker flag that exposes port 19891 for debugging the Lambda handler code that will run inside the container.
+
+You will also need to install our LocalStack AWS CLI, which is a thin wrapper around the aws command line interface for use with LocalStack, using `pip`:
+
+```sh
+pip install awscli-local
+```
+
+## Configuring the code for remote debugging
+
+After setting up LocalStack and launching it in detached mode, we can now configure the debug server using `debugpy`. `debugpy` is an implementation of the [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/) for Python 3 and is used for debugging on VS Code. `debugpy` implements all of the standard debugging tools you would expect and provides a standardised way for development tools to communicate with debuggers.
+
+To enable `debugpy` inside the Lambda function, you need to place the following block of code inside the handler code:
+
+```python
+import debugpy 
+debugpy.listen(19891) 
+debugpy.wait_for_client()
+```
+
+The `wait_for_client()` function blocks execution until client is attached. You can also use the `wait_for_debug_client` function which implements the start of the debug server and also adds an automatic cancellation of the wait task if the debug client (i.e. VSCode) doesn’t connect.
+
```

Comment:
We should probably mention why we're extending `sys.path` in the snippet below - the main reason we added it in the Pro sample and in the screenshot was to speed up the Lambda startup cycle.

Something like:
> Please note that in the code snippet below we assume that the `debugpy` package has been installed into a virtualenv folder named `.venv` inside the same directory where the Lambda handler (`handler.py`) is stored on the local disk (see further below). Hence, adding `.venv/lib/python*/site-packages` to the Python system path ensures that `debugpy` is available to the Lambda handler execution at runtime.